src/pipelines/pipeline.py

```diff
@@ -1,27 +1,112 @@
-import os
+import gc
+import logging
+import time
 from argparse import Namespace
-from typing import List, Tuple, Union
+from typing import Any, Dict, List, Tuple, Type
 
+import numpy as np
 import torch
 
-from transformers import AutoTokenizer, BloomConfig, BloomForCausalLM, GPT2Config, GPT2LMHeadModel
+from src.utils.arguments import check_unused
+from src.utils.fast_init import fast_init
+from src.utils.logging import format_ms, log_rank_n
+from transformers import AutoTokenizer, BloomForCausalLM, GPT2LMHeadModel, PretrainedConfig, PreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+NUM_GENERATED_TOKENS = "num_generated_tokens"
+TOKENIZE_TIME = "tokenize_time"
+MODEL_TIME = "model_time"
+DECODE_TIME = "decode_time"
+END_TO_END_TIME = "end_to_end_time"
+
+METRIC_KEYS = (
+    NUM_GENERATED_TOKENS,
+    TOKENIZE_TIME,
+    MODEL_TIME,
+    DECODE_TIME,
+    END_TO_END_TIME,
+)
 
 
 class Pipeline:
     def __init__(self, args: Namespace) -> None:
-        self.config, self.tokenizer, self.model_class = get_config_tokenizer_model_class(args)
-        self.model = None
-        self.input_device = None
-
-    def __call__(self, text: List[str], **generate_kwargs) -> Tuple[List[str], List[int]]:
+        log_rank_n("*** Setting up tokenizer", logger.info)
+        self.tokenizer = AutoTokenizer.from_pretrained("gpt2")
+        self.tokenizer.add_special_tokens({"pad_token": "[PAD]"})
+
+        self.device = args.device
+
+        model_class, config = self.get_config(args)
+        is_int8 = args.dtype == torch.int8
+        if is_int8:
+            check_unused(args, {"device": torch.device("cuda")}, enforce=True)
+        torch_dtype = torch.float16 if is_int8 else args.dtype
```

Comment:
shouldn't this be torch.int8?

Reply:
We can't create the model directly in int8, we have to go through a float.